### ISSUE TITLE
Your eyes can deceive you...

### ISSUE THREAD

**Issue by I-am-Erk** [CleverRaven/Cataclysm-DDA#53171]
### Is your feature request related to a problem? Please describe.

Kaluptic psychosis. It's awesome, isn't it? Some of it is obviously off (at least that's what my multi-cooker tells me), but who hasn't had the fight of their life against a zombie that turned out not to exist.

Messing with the player's perception of the world through their avatar is a powerful and really interesting play experience, and I think it's one we need more of. Everything in this post can, and should, be considered as an option for kaluptic psychosis effects... but it should also be made available to other perception-altering events and occasions, such as portal storms or as [side effects for injecting too many experimental drugs all at once](#53141), or even just as a way of representing poor vision.

### Solution you would like.

These are each separate ideas for ways we could mess with the player's display. Most of these require a fundamental feature that may need code support first: we need to be able to have some items that block the affected character's line of sight but not other actors. I think we already have code for that in some cases, like mirrors, but I am not sure if it's extensible here or not. Kevin suggests: "We might be able to just have a secondary FoV output for hallucinatory FoV for just when that kind of hallucination is active. It's pretty simple to do, you just need to edit the transparency mask you pass to FoV calc."

Ideally code should account for the possibility that NPCs see this stuff, so like, pathing should also take into account imagined terrain. Ideally also, map memory should store what you think is in the tile, not what really is in it.

1. Mirages:
A mirage would be a map extra that can be placed anywhere. It can only contain terrain and furniture. Only the affected character can see it. If you move into terrain occupied by the mirage, the whole thing vanishes. loo`k`ing at the squares occupied by the mirage gives you the usual response for the terrain and furiture contained therein. Monsters and items render normally in the space, so you may see zombies walk through an illusory wall. Mirages should be defined exactly like map extras, so that certain mirages might appear only in certain environments, and existing map extras can be used as mirages.

A mirage should appear on map memory when it generates. This may change your map memory of an area, so you might go to a familiar place and see a shed in the map memory and in "reality", until you walk up to it.

2. Terrain scramble:
Starting at distance from character _n_, based on intensity: a random chance that a perceived object (monster, furniture, terrain, or item) might be swapped with an adjacent object at rendering time. As distance from character increases, the chance of sprites being swapped increases, so your information about further away tiles is less and less accurate.
There should be an algorithm to keep this from producing meaningless scramble:
- don't swap identical objects.
- if a sprite has CONNECTS_TO as a field, check if it is currently connecting to anything as a multitile. If so, it can only swap with a sprite it is connected to. That prevents walls from swapping with grass, which would be an obvious hallucination.
- If the sprites have assymetric line-of-sight blocking properties (eg you swapped a tree with a shrub) use the most line-of-sight blocking properties for both (you can't see through the tree's actual square, nor through the shrub square that is being rendered as a tree).
- we might have a parallel non-identical function that allows vehicles in the distance to be perceived in slightly different locations. This would move the whole perceived vehicle, not just one tile.

If done well, this could also be used to represent nearsightedness, blurred vision, and trying to make stuff out in the dark.

3. Misremembering:
Similar to mirages, we should allow individual sprites to be plunked wholesale into your map memory. If we're able, these remembered sprites should be allowed to include things that can't normally appear in map memory, so you could see zombies and things appearing. Perhaps we'd even allow some novel sprites that only show up there. "Misremembered" hallucinations should have context for what terrain they can appear on top of, so that we can eg. have forests appear alien and spooky in your map memory, and so that monster memories only appear on floor tiles and not hovering over walls.

### Describe alternatives you have considered.

Not doing this.

### Additional context

I'd like to really mess with people's heads when they take weird meds, yknow?

**Comment by ERRORCODE509** [comment]
What about also messing with the player's perception of the overmap? I think that could be interesting.

**Comment by I-am-Erk** [comment]
I see no reason why not, what would you suggest?

**Comment by ERRORCODE509** [comment]
Perhaps it could occasionally change the names and appearances of overmap tiles outside your overmap view distance. It would probably work better for cities, as making a house appear to be a gun store or something could make things a little more challenging. Plains and forests are a bit harder, maybe migo scout towers, LMOE shelters, caves, that sort of thing.

**Comment by PatrikLundell** [comment]
Aren't mirages physical phenomena caused by light reflection (which then interacts with the brain's attempt to interpret them)? In so, they should affect others with the same point of view (you can point out "puddles" on/over hot asphalt to someone next to you, for instace). What's described seems to be a hallucination effect.

**Comment by I-am-Erk** [comment]
Just reusing the name for a vaguely similar effect, because there are only so many words in english and "hallucinatory terrain" is a mouthful. Plus, like a mirage, it vanishes as you approach.

**Comment by Hirmuolio** [comment]
Light level visibility mechanic could also be tied to this.

Currently there are three levels: Normal (full vision), dim (dark overlay but still full vision) and dark (no vision).
Objects in the dim area could be replaced with ambiguous "there is something here" marker. It could differentiate between sizes and movement but wouldn't show exact info. And sometimes you could see the thing in the darkness as something else.

**Comment by I-am-Erk** [comment]
> Light level visibility mechanic could also be tied to this.
> 
> Currently there are three levels: Normal (full vision), dim (dark overlay but still full vision) and dark (no vision). Objects in the dim area could be replaced with ambiguous "there is something here" marker. It could differentiate between sizes and movement but wouldn't show exact info. And sometimes you could see the thing in the darkness as something else.

That is a neat idea

**Comment by estebandellasilva** [comment]
Maybe wrong or misinterpreting this issue, but in most popular media, when a character iss hot/thirst/tired they are more likely to see things which arent there, would this apply here?

**Comment by I-am-Erk** [comment]
Not really. There are some hallucinations you can experience in extremis but this isn't well suited to represent them

**Comment by MNG-cataclysm** [comment]
What about auditory hallucinations? Shuffling in the distance when nothing is there, or someone calling to you while you are alone. The sounds of certain monsters could play at random, possibly coupled with a visual hallucination upon investigation.

**Comment by I-am-Erk** [comment]
Yeah that's a great idea
